tests/native/test_syscalls.py

```diff
@@ -141,6 +142,56 @@ def test_chmod(self):
 
         self.assertEqual(-errno.EPERM, self.linux.sys_chown(0x1100, 0, 0))
 
+    def test_recvfrom(self):
+        self.linux.current.memory.mmap(0x1000, 0x1000, "rw ")
+
+        sock_fd = self.linux.sys_socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+        self.assertEqual(sock_fd, 3)
+        # Unimplemented
+        # self.linux.current.write_int(0x1000, 1, size=8 * 4)
+        # self.linux.sys_setsockopt(sock_fd, socket.SOL_SOCKET, socket.SO_REUSEPORT, 0x1000, 4)
+        self.linux.sys_bind(sock_fd, None, None)
+        self.linux.sys_listen(sock_fd, None)
+        conn_fd = self.linux.sys_accept(sock_fd, None, 0)
+        self.assertEqual(conn_fd, 4)
+
+        sock_obj = self.linux.files[conn_fd]
+        init_len = len(sock_obj.buffer)
+        BYTES = 5
+        wrote = self.linux.sys_recvfrom(conn_fd, 0x1100, BYTES, 0x0, 0x0, 0x0)
+        self.assertEqual(wrote, BYTES)
+
+        wrote = self.linux.sys_recvfrom(conn_fd, 0x0, 100, 0x0, 0x0, 0x0)
+        self.assertEqual(wrote, -errno.EFAULT)
+
+        remain_len = init_len - BYTES
+        self.assertEqual(remain_len, len(sock_obj.buffer))
+
+        wrote = self.linux.sys_recvfrom(conn_fd, 0x1100, remain_len + 10, 0x0, 0x0, 0x0)
+        self.assertEqual(wrote, remain_len)
+
+        wrote = self.linux.sys_recvfrom(conn_fd, 0x1100, 10, 0x0, 0x0, 0x0)
+        self.assertEqual(wrote, 0)
+
+        self.linux.sys_close(conn_fd)
+        wrote = self.linux.sys_recvfrom(conn_fd, 0x1100, 10, 0x0, 0x0, 0x0)
+        self.assertEqual(wrote, -errno.EINVAL)
+
+    @unittest.expectedFailure
+    def test_multiple_sockets(self):
+        sock_fd = self.linux.sys_socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+        self.assertEqual(sock_fd, 3)
+        self.linux.sys_bind(sock_fd, None, None)
+        self.linux.sys_listen(sock_fd, None)
+        conn_fd = self.linux.sys_accept(sock_fd, None, 0)
+        self.assertEqual(conn_fd, 4)
+        self.linux.sys_close(conn_fd)
+
+        conn_fd = -1
+        # Fails with "Name socket4 already in use"
```

Comment:
Since we've closed FD 4 already, shouldn't this be okay? The tests pass so is this just an outdated comment?

Reply:
We did close FD 4, but `sys_accept` should be able to either open a new FD number or re-use FD 4 (native run re-uses 4). Currently, it tries to reopen FD 4, but we haven't purged FD 4 from the available new FDs for open files list when we close it (or something, I forget exactly), so there is an error when trying open a new FD after closing.

Here is a C program that might be a bit more concrete:

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <netinet/in.h>
#include <signal.h>
#include <string.h>

#define PORT 8081
#define BUFFER_SIZE 80
#define TRUE 0
#define FALSE 1

int server_fd, new_client;
struct sockaddr_in address;

int setupServer()
{
    int opt = 1; // reuse address

    // create socket file descriptor, attach to 8081
    server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) {
      printf("Cannot setsockopt\n");
      return FALSE;
    }
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons(PORT);
    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address))) {
      printf("Cannot bind\n");
      return FALSE;
    }

    printf("Listening on port %i...\n", PORT);
    return TRUE;
}

int main() {
  int setup = setupServer();
  if (setup != 0) {
    return 1;
  }

  int addrlen = sizeof(address);

  listen(server_fd, 10);
  printf("Listening on fd %d...\n", server_fd);
  new_client = accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen);
  printf("Accepting on fd %d...\n", new_client);
  close(new_client);
  new_client = accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen);
  printf("Accepting on fd %d...\n", new_client);
  close(new_client);
  close(server_fd);
}
```

Some manticore output (from this branch)

```
...
2019-09-03 21:40:56,816: [47] m.p.platform:DEBUG: sys_arch_prctl(0x1002, 0x7ffffffd5f00) = 0
2019-09-03 21:41:20,429: [47] m.p.platform:DEBUG: sys_mprotect(0x7fffffdab000, 0x4000, 1) = 0
2019-09-03 21:41:24,487: [47] m.p.platform:DEBUG: sys_mprotect(0x555555754000, 0x1000, 1) = 0
2019-09-03 21:41:26,867: [47] m.p.platform:DEBUG: sys_mprotect(0x7ffffffdc000, 0x1000, 1) = 0
2019-09-03 21:41:26,900: [47] m.p.platform:DEBUG: sys_munmap(0x7ffffffd7000, 0x406e) = 0
2019-09-03 21:41:28,207: [47] m.p.platform:DEBUG: sys_socket(2, 1, 0) = 3
2019-09-03 21:41:28,222: [47] m.p.platform:WARNING: Unimplemented system call at 0x7fffffae6e6a: sys_setsockopt(0x3, 0x1, 0xf, 0x7ffffffffdb4, 0x4)
2019-09-03 21:41:28,222: [47] m.p.platform:DEBUG: sys_setsockopt(3, 1, 15, "", 4) = 0
2019-09-03 21:41:28,242: [47] m.p.platform:DEBUG: sys_bind(3, "", 16) = 0
2019-09-03 21:41:28,471: [47] m.p.platform:DEBUG: sys_newfstat(1, 0x7ffffffff600) = 0
2019-09-03 21:41:28,508: [47] m.p.platform:DEBUG: sys_ioctl(1, 0x5401, 0x7ffffffff560) = -22
2019-09-03 21:41:28,598: [47] m.n.c.abstractcpu:WARNING: Unimplemented instruction: 0x00007fffffdcc631: 0f ae 44 24 40  fxsave [rsp + 0x40]
2019-09-03 21:41:28,972: [47] m.n.c.abstractcpu:WARNING: Unimplemented instruction: 0x00007fffffdcc646: 0f ae 4c 24 40  fxrstor[rsp + 0x40]
2019-09-03 21:41:49,827: [47] m.n.c.abstractcpu:WARNING: Unimplemented instruction: 0x00007fffffdcc631: 0f ae 44 24 40  fxsave [rsp + 0x40]
2019-09-03 21:41:50,031: [47] m.n.c.abstractcpu:WARNING: Unimplemented instruction: 0x00007fffffdcc646: 0f ae 4c 24 40  fxrstor[rsp + 0x40]
2019-09-03 21:41:50,309: [47] m.p.platform:DEBUG: sys_brk(0) = 0x555555756000
2019-09-03 21:41:50,330: [47] m.p.platform:DEBUG: sys_brk(0x555555777000) = 0x555555777000
2019-09-03 21:41:51,237: [47] m.p.platform:DEBUG: sys_listen(3, 10) = 0
2019-09-03 21:41:51,531: [47] m.p.platform:DEBUG: sys_accept(3, "", "") = 4
2019-09-03 21:41:51,832: [47] m.p.linux:DEBUG: sys_close(4)
2019-09-03 21:41:51,832: [47] m.p.platform:DEBUG: sys_close(4) = 0
2019-09-03 21:41:51,846: [47] m.c.worker:ERROR: Exception in state 1: ValueError('Name socket4 already used',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/manticore/platforms/linux.py", line 2420, in execute
    self.current.execute()
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/abstractcpu.py", line 1004, in execute
    raise e
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/abstractcpu.py", line 990, in execute
    implementation(*insn.operands)
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/abstractcpu.py", line 1133, in new_method
    return old_method(cpu, *args, **kw_args)
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/x86.py", line 5663, in SYSCALL
    raise Syscall()
manticore.native.cpu.abstractcpu.Syscall: CPU Syscall

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/manticore/core/worker.py", line 121, in run
    current_state.execute()
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/state.py", line 29, in execute
    result = self._platform.execute()
  File "/usr/local/lib/python3.6/dist-packages/manticore/platforms/linux.py", line 2427, in execute
    self.syscall()
  File "/usr/local/lib/python3.6/dist-packages/manticore/platforms/linux.py", line 2254, in syscall
    return self._syscall_abi.invoke(implementation)
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/abstractcpu.py", line 425, in invoke
    ret = super().invoke(model, prefix_args)
  File "/usr/local/lib/python3.6/dist-packages/manticore/native/cpu/abstractcpu.py", line 367, in invoke
    result = model(*arguments)
  File "/usr/local/lib/python3.6/dist-packages/manticore/platforms/linux.py", line 3013, in sys_accept
    symb = self.constraints.new_array(name=f"socket{fd}", index_max=nbytes)
  File "/usr/local/lib/python3.6/dist-packages/manticore/core/smtlib/constraints.py", line 397, in new_array
    raise ValueError(f"Name {name} already used")
ValueError: Name socket4 already used

2019-09-03 21:41:52,015: [17] m.c.plugin:INFO: Instructions executed: 0
2019-09-03 21:41:52,023: [17] m.c.plugin:INFO: Coverage: 0 different instructions executed
2019-09-03 21:41:52,087: [17] m.c.manticore:INFO: Generated testcase No. 0 - test
2019-09-03 21:41:52,551: [17] m.c.manticore:INFO: Results in /test/mcore_1yzdufl9
2019-09-03 21:41:52,552: [17] m.n.manticore:INFO: Total time: 62.54935026168823
```

You can run it natively and interact with it by just issuing `netcat 127.0.0.1 8081` a couple times to pass the `accept`s and it should print:
```
root@b217dbd5198c:/test# ./a.out
Listening on port 8081...
Listening on fd 3...
Accepting on fd 4...
Accepting on fd 4...
```

Reply:
Oh, this is a problem with smtlib. We free up FD 4 for the creation of a new socket, but we don't delete the backing symbolic array "socket4" from the constraint set, so when we try to create a new socket, it yells at us. Previously this was even worse, as we'd use the name "socket" for every backing array so you couldn't make more than one. 

If you want to fix it, you could do one of the following:
* Count the symbols we've created with a given name and deduplicate them by postfixing the counter
* Replace the file descriptor number in the symbol name with a long-ish random number
* Change the name of the symbol in the constraint set as soon as the FD gets closed.

Reply:
Ah. That makes more sense. Thank you!

Alos, if it's okay with you, I think a different PR to fix this overlapping symbolic array issue would be more appropriate than including a fix in this PR.

I've opened #1521 to keep track of this.